src/Compilers/Core/MSBuildTask/GenerateMSBuildAnalyzerConfig.cs

```diff
@@ -0,0 +1,107 @@
+    /// inlcuded the file (e.g. <c>Compile</c>, <c>AdditionalFile</c> etc.); <c>MetadataName</c> is expected to contain the name of
+    /// another piece of metadata that should be retreived and used as the output value in the file. It is expected that a given 
+    /// file can have multiple entries in the <see cref="MetadataItems" /> differing by its <c>ItemType</c>.
+    /// 
+    /// Each of the <see cref="MetadataItems"/> will be transformed into a new section in the generated config file. The section
+    /// header will be the full path of the item (generated via its<see cref="ITaskItem.ItemSpec"/>), and each section will have a 
+    /// set of <c>msbuild_metadata.<em>ItemType</em>.<em>MetadataName</em> = <em>RetreivedMetadataValue</em></c>, one per <c>ItemType</c>
+    /// 
+    /// The Microsoft.Managed.Core.targets calls this task with the collected results of the <c>AnalyzerProperty</c> and 
+    /// <c>AnalyzerItemMetadata</c> item groups. 
+    /// </remarks>
+    public sealed class GenerateMSBuildAnalyzerConfig : Task
+    {
+        [Output]
+        public string ConfigFileContents { get; set; }
+
+        [Required]
+        public ITaskItem[] MetadataItems { get; set; }
+
+        [Required]
+        public ITaskItem[] PropertyItems { get; set; }
+
+        public GenerateMSBuildAnalyzerConfig()
+        {
+            ConfigFileContents = string.Empty;
+            MetadataItems = Array.Empty<ITaskItem>();
+            PropertyItems = Array.Empty<ITaskItem>();
+        }
+
+        public override bool Execute()
+        {
+            StringBuilder builder = new StringBuilder();
+
+            // we always generate global configs
+            builder.AppendLine("is_global = true");
+
+            // collect the properties into a global section
+            foreach (var prop in PropertyItems)
+            {
+                builder.Append("msbuild_property.")
+                       .Append(prop.ItemSpec)
+                       .Append(" = ")
+                       .AppendLine(prop.GetMetadata("Value"));
+            }
+
+            // group the metadata items by their full path
+            var groupedItems = MetadataItems.GroupBy(i => NormalizeWithForwardSlash(i.GetMetadata("FullPath")));
+
+            foreach (var group in groupedItems)
+            {
+                // write the section for this item
+                builder.AppendLine()
+                       .Append("[")
+                       .Append(group.Key)
+                       .AppendLine("]");
+
+                foreach (var item in group)
+                {
+                    string metadataName = item.GetMetadata("MetadataName");
+                    builder.Append("msbuild_item.")
+                           .Append(item.GetMetadata("ItemType"))
+                           .Append(".")
+                           .Append(metadataName)
```

Comment:
Will `metadataName` always be a valid identifier for a editor config value name here?

Reply:
Yes. Editorconfig keys are opaque strings; the only disallowed characters are whitespace and '=' (see [https://editorconfig-specification.readthedocs.io/en/latest/#id3](https://editorconfig-specification.readthedocs.io/en/latest/#id3)).  

`metadataName` is the value of either an xml element or attribute name (note: the name, not it's value). The XML spec disallows '=' or ' ' as identifier characters ([https://www.w3.org/TR/2008/REC-xml-20081126/#NT-Name](https://www.w3.org/TR/2008/REC-xml-20081126/#NT-Name)) which means it will always be valid.

---
In reply to: [419844440](https://github.com/dotnet/roslyn/pull/43617#discussion_r419844440) [](ancestors = 419844440)